src/jobserver.cc

```diff
@@ -0,0 +1,297 @@
+        config->mode = Jobserver::Config::kModePosixFifo;
+        config->path = fifo_path.AsString();
+      } else {
+        config->mode = Jobserver::Config::kModeWin32Semaphore;
+        config->path = value.AsString();
+      }
+      continue;
+    }
+
+    // Handle --jobserver-fds which is an old undocumented variant of
+    // --jobserver-auth that only accepts a pair of file descriptor.
+    // This was replaced by --jobserver-auth=R,W in GNU Make 4.2.
+    if (GetPrefixedValue(arg, "--jobserver-fds=", &value)) {
+      if (!GetFileDescriptorPair(value, config)) {
+        *error = "Invalid file descriptor pair [" + value.AsString() + "]";
+        return false;
+      }
+      config->mode = Jobserver::Config::kModePipe;
+      continue;
+    }
+
+    // Ignore this argument. This assumes that MAKEFLAGS does not
+    // use spaces to separate the option from its argument, e.g.
+    // `--jobserver-auth <something>`, which has been confirmed with
+    // Make 4.3, even if it receives such a value in its own env.
+  }
+
+  return true;
+}
+
+bool Jobserver::ParseNativeMakeFlagsValue(const char* makeflags_env,
+                                          Jobserver::Config* config,
+                                          std::string* error) {
+  if (!ParseMakeFlagsValue(makeflags_env, config, error))
+    return false;
+
+#ifdef _WIN32
+  if (config->mode == Jobserver::Config::kModePosixFifo) {
+    *error = "FIFO mode is not available on Windows!";
+    return false;
+  }
+  if (config->mode == Jobserver::Config::kModePipe) {
+    *error =
+        "File-descriptor based authentication is not available on Windows!";
+    return false;
+  }
+#else   // !_WIN32
+  if (config->mode == Jobserver::Config::kModeWin32Semaphore) {
+    *error = "Semaphore mode is only available on Windows!";
+    return false;
+  }
+#endif  // !_WIN32
+  return true;
+}
+
+// static
+std::string Jobserver::Config::ModeToString(Jobserver::Config::Mode mode) {
+  std::string result;
+  for (const auto& valid : kValidModes) {
+    if (valid.mode_value == mode) {
+      result = valid.mode_str;
+      break;
+    }
+  }
+  return result;
+}
+
+std::pair<bool, Jobserver::Config::Mode> Jobserver::Config::ModeFromString(
+    const std::string& mode_str) {
+  std::pair<bool, Jobserver::Config::Mode> result = {
+    false, Jobserver::Config::kModeNone
+  };
+  for (const auto& valid : kValidModes) {
+    if (mode_str == valid.mode_str) {
+      result.first = true;
+      result.second = valid.mode_value;
+      break;
+    }
+  }
+  return result;
+}
+
+std::string Jobserver::Config::GetValidModesListAsString(
+    const char* separator) {
+  std::string result;
+  for (const auto& valid : kValidModes) {
+    if (!result.empty())
+      result += separator;
+    result += valid.mode_str;
+  }
+  return result;
+}
+
+#ifdef _WIN32
+#include "jobserver-win32.h"
+#else  // !_WIN32
+#include "jobserver-posix.h"
+#endif  // !_WIN32
```

Comment:
Name the files `.cc` or `.inc.cc` to indicate that they are to be included by a `.cc` file.

Reply:
Yes, this was a bit unsavory. I wanted to avoid changing the CMakeLists.txt and configure.py files with conditional paths. I refactored these into proper `.cc` files each, with no extra includes, which actually allowed for some simplification after the fact. Thanks for pointing this out.